### PR TITLE
Add TLSAcceptor and Builder

### PR DESCRIPTION
**Pull request by Licenser**
TLSAcceptor, including a builder, also adds a `builder()` method to the Connector for convenience.

### CODE REVIEW

**Comment by Licenser** [comment]
I'm stumped, the CI errors with:

```
error[E0432]: unresolved imports `hyper::server::conn::AddrIncoming`, `hyper::server::conn::AddrStream`
 --> src/acceptor.rs:5:12
  |
5 |     conn::{AddrIncoming, AddrStream},
  |            ^^^^^^^^^^^^  ^^^^^^^^^^ no `AddrStream` in `server::conn`
  |            |
  |            no `AddrIncoming` in `server::conn`
```

the `server` feature is part of the `Cargo.toml` and locally it works.

[AddrIncoming](https://docs.rs/hyper/latest/hyper/server/conn/struct.AddrIncoming.html) requires just the `server` feature regarding to the docs too so I honestly don't know why the CI hates me :(

**Comment by Licenser** [comment]
This is what happens locally for me and I can't find the difference between my box and the CI
```
❯ cargo test --no-default-features
   Compiling rustls v0.20.8
   Compiling tokio-rustls v0.23.4
   Compiling hyper-rustls v0.23.2 (/home/heinz/hyper-rustls)
warning: unused import: `rustls::client::WantsTransparencyPolicyOrClientCert`
 --> src/config.rs:1:5
  |
1 | use rustls::client::WantsTransparencyPolicyOrClientCert;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: unused macro definition: `trace`
  --> src/lib.rs:90:18
   |
90 |     macro_rules! trace    ( ($($tt:tt)*) => {{}} );
   |                  ^^^^^
   |
   = note: `#[warn(unused_macros)]` on by default

warning: unused macro definition: `debug`
  --> src/lib.rs:91:18
   |
91 |     macro_rules! debug    ( ($($tt:tt)*) => {{}} );
   |                  ^^^^^

warning: unused imports: `debug`, `trace`
  --> src/lib.rs:92:21
   |
92 |     pub(crate) use {debug, trace};
   |                     ^^^^^  ^^^^^

warning: `hyper-rustls` (lib) generated 4 warnings
warning: unused import: `super::ConnectorBuilder as HttpsConnectorBuilder`
   --> src/connector/builder.rs:264:9
    |
264 |     use super::ConnectorBuilder as HttpsConnectorBuilder;
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: `hyper-rustls` (lib test) generated 5 warnings (4 duplicates)
    Finished test [unoptimized + debuginfo] target(s) in 4.03s
     Running unittests src/lib.rs (target/debug/deps/hyper_rustls-5f605abbeb6420e6)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests/tests.rs (target/debug/deps/tests-649482a5c3748cb9)

running 3 tests
Starting to serve on https://127.0.0.1:1338.
Starting to serve on https://127.0.0.1:1337.
test client ... ok
test server ... ok
test custom_ca_store ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.02s

   Doc-tests hyper-rustls

running 3 tests
test src/lib.rs - (line 8) - compile ... ok
test src/lib.rs - (line 32) - compile ... ok
test src/connector/builder.rs - connector::builder::ConnectorBuilder (line 17) ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.33s
```

**Comment by Licenser** [comment]
@djc when you have some time I could use a extra pairs of eyes since the errors in CI make no sense to me :(

**Comment by Licenser** [comment]
I may or may not have solved the CI issue, I made a feature flag for the server as it pulls in additional (sub) dependencies otherwise and I wanted to avoid that as default

File: examples/server.rs [hunk 1]
```diff
@@ -4,20 +4,18 @@
...
 use std::vec::Vec;
-use std::{env, fs, io, sync};
-use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
-use tokio_rustls::rustls::ServerConfig;
+use std::{env, fs, io};
 
+#[cfg(not(feature = "acceptor"))]
+fn main() {
+    println!("This example requires the `acceptor` feature");
+}
```
Comment (djc) [inline examples/server.rs:17]: I don't think this is helpful. How about we just stick a `#![cfg(feature = "acceptor")]` at the top of the module?

File: src/acceptor.rs [hunk 2]
```diff
@@ -0,0 +1,139 @@
+use core::task::{Context, Poll};
```
Comment (djc) [inline src/acceptor.rs:1]: Please split this in two blocks: `core` and `std` imports and other external crate imports.

File: src/acceptor.rs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+use std::io;
+use std::pin::Pin;
+use std::sync::Arc;
+use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
+
+mod builder;
+
+pub use builder::AcceptorBuilder;
+
+use self::builder::WantsTlsConfig;
```
Comment (djc) [inline src/acceptor.rs:18]:
```suggestion
mod builder;
pub use builder::AcceptorBuilder;
use builder::WantsTlsConfig;

```

File: src/acceptor/builder.rs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+pub struct AcceptorBuilder<State>(State);
+
+/// State of a builder that needs a TLS client config next
+pub struct WantsTlsConfig(());
+
+impl AcceptorBuilder<WantsTlsConfig> {
+    /// Creates a new [`AcceptorBuilder`]
+    pub fn new() -> Self {
+        Self(WantsTlsConfig(()))
+    }
```
Comment (djc) [inline src/acceptor/builder.rs:17]: Let's make this private for now, so that the only way to construct it is through `Acceptor::builder()`.

File: src/acceptor/builder.rs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        Ok(AcceptorBuilder(WantsAlpn(
+            ServerConfig::builder()
+                .with_safe_defaults()
+                .with_no_client_auth()
+                .with_single_cert(cert_chain, key_der)?,
+        )))
+    }
+}
+
+impl Default for AcceptorBuilder<WantsTlsConfig> {
```
Comment (djc) [inline src/acceptor/builder.rs:45]: I would prefer to remove this.

File: src/acceptor/builder.rs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    /// Configure ALPN to accep HTTP/1.1
+    pub fn with_http11_alpn(mut self) -> AcceptorBuilder<WantsIncoming> {
+        self.0 .0.alpn_protocols = vec![b"http/1.1".to_vec()];
+        AcceptorBuilder(WantsIncoming(self.0 .0))
+    }
+
+    /// Configure ALPN to accept HTTP/2, HTTP/1.1 in that order.
+    pub fn with_http_alpn(mut self) -> AcceptorBuilder<WantsIncoming> {
```
Comment (djc) [inline src/acceptor/builder.rs:77]: Naming nit: this is a little implicit. It just says `http` but it's actually HTTP/2 and HTTP/1.1. Once hyper supports H3, should we expect this to also support it? I think it would be good to mirror the `ConnectorBuilder` methods here, and have `with_http1_alpn()`, `with_http2_alpn()`, `with_all_versions_alpn()`.

File: src/connector.rs [hunk 7]
```diff
@@ -31,6 +31,13 @@ impl<T> fmt::Debug for HttpsConnector<T> {
     }
 }
 
+impl HttpsConnector<()> {
```
Comment (djc) [inline src/connector.rs:34]: This seems unrelated to the acceptor changes? I'd prefer to have this PR focused on the acceptor.

File: src/lib.rs [hunk 8]
```diff
@@ -26,10 +26,57 @@
...
+//! # #[cfg(all(feature = "rustls-native-certs", feature = "tokio-runtime", feature = "http1", feature = "acceptor"))]
+//! # fn main() {
+//! use hyper::server::conn::AddrIncoming;
+//! use hyper::service::{make_service_fn, service_fn};
+//! use hyper::{Body, Method, Request, Response, Server, StatusCode};
+//! use hyper_rustls::TlsAcceptor;
+//! use std::io;
+//! use std::fs::File;
+//!
+//! let mut rt = tokio::runtime::Runtime::new().unwrap();
```
Comment (djc) [inline src/lib.rs:42]: Would probably be good to add some an empty line before each comment to make this a bit easier to consume.

**Review by djc — COMMENTED** [review]
Thanks for working on this, and sorry for the slow follow-up!

Would be great to get this over the finish line.

File: src/connector.rs (hunk 7 above)
Comment (djc) [reply src/connector.rs:43]: Fine, let's keep the `AcceptorBuilder::new()`. I still don't think we should add this here, since the `()` type argument makes this a little weird and unergonomic.

File: Cargo.toml [hunk 9]
```diff
@@ -19,21 +19,28 @@ rustls = { version = "0.21.0", default-features = false }
...
 webpki-roots = { version = "0.23", optional = true }
+futures-util = { version = "0.3" }
 
 [dev-dependencies]
 futures-util = { version = "0.3.1", default-features = false }
 hyper = { version = "0.14", features = ["full"] }
 rustls = { version = "0.21.0", default-features = false, features = ["tls12"] }
 rustls-pemfile = "1.0.0"
-tokio = { version = "1.0", features = ["io-std", "macros", "net", "rt-multi-thread"] }
+tokio = { version = "1.0", features = [
```
Comment (djc) [inline Cargo.toml:29]: This is a formatting only change, please leave it out since it's irrelevant to the other changes.

**Comment by Licenser** [comment]
Should be all done :)

File: src/acceptor/builder.rs [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    /// Configure ALPN to accept HTTP/2, HTTP/1.1 in that order.
+    pub fn with_all_versions_alpn(mut self) -> AcceptorBuilder<WantsIncoming> {
+        self.0 .0.alpn_protocols = vec![b"h2".to_vec(), b"http/1.1".to_vec()];
+        AcceptorBuilder(WantsIncoming(self.0 .0))
+    }
+}
+
+/// State of a builder that needs a incoming address next
+pub struct WantsIncoming(ServerConfig);
+impl AcceptorBuilder<WantsIncoming> {
```
Comment (djc) [inline src/acceptor/builder.rs:79]:
```suggestion
pub struct WantsIncoming(ServerConfig);

impl AcceptorBuilder<WantsIncoming> {
```

File: src/acceptor/builder.rs [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    /// Passes a [`AddrIncoming`] to configure the TLS connection and
+    /// creates the [`TlsAcceptor`]
+    pub fn with_incoming<I>(self, incoming: I) -> TlsAcceptor
+    where
+        AddrIncoming: From<I>,
```
Comment (djc) [inline src/acceptor/builder.rs:84]:
```suggestion
    pub fn with_incoming(self, incoming: impl Into<AddrIncoming>) -> TlsAcceptor {```

Comment (djc) [reply src/acceptor/builder.rs:84]: Actually, it looks like `AddrIncoming` doesn't have any `From` impls, so we can just use the concrete type here?

File: src/lib.rs [hunk 12]
```diff
@@ -3,7 +3,7 @@
 //! A pure-Rust HTTPS connector for [hyper](https://hyper.rs), based on
 //! [Rustls](https://github.com/rustls/rustls).
 //!
-//! ## Example
+//! ## Example Client
```
Comment (djc) [inline src/lib.rs:6]:
```suggestion
//! ## Example client
```

File: src/lib.rs [hunk 13]
```diff
@@ -26,10 +26,65 @@
...
+//! use hyper::service::{make_service_fn, service_fn};
+//! use hyper::{Body, Method, Request, Response, Server, StatusCode};
+//! use hyper_rustls::TlsAcceptor;
+//! use std::io;
+//! use std::fs::File;
+//!
+//! let mut rt = tokio::runtime::Runtime::new().unwrap();
+//! let addr = "127.0.0.1:1337".parse().unwrap();
+//!
+//!
```
Comment (djc) [inline src/lib.rs:45]:
```suggestion
//!
```

File: src/lib.rs [hunk 14]
```diff
@@ -26,10 +26,65 @@
...
+//! // Load public certificate.
+//! let certfile = File::open("examples/sample.pem").unwrap();
+//! let mut reader = io::BufReader::new(certfile);
+//!
+//!
```
Comment (djc) [inline src/lib.rs:50]:
```suggestion
//!
```

File: src/lib.rs [hunk 15]
```diff
@@ -26,10 +26,65 @@
...
+//! // Load and return certificate.
+//! let certs = rustls_pemfile::certs(&mut reader).unwrap();
+//! let certs =certs.into_iter().map(rustls::Certificate).collect();
+//!
+//!
```
Comment (djc) [inline src/lib.rs:55]:
```suggestion
//! let certs = certs.into_iter().map(rustls::Certificate).collect();
//!
```

File: src/lib.rs [hunk 16]
```diff
@@ -26,10 +26,65 @@
...
+//! // Load private key. (see `examples/server.rs`)
+//! let keyfile = File::open("examples/sample.rsa").unwrap();
+//! let mut reader = io::BufReader::new(keyfile);
+//!
+//!
```
Comment (djc) [inline src/lib.rs:60]:
```suggestion
//!
```

File: src/acceptor.rs [hunk 17]
```diff
@@ -0,0 +1,139 @@
+use core::task::{Context, Poll};
+use std::future::Future;
+use std::io;
+use std::pin::Pin;
+use std::sync::Arc;
+
+use futures_util::ready;
```
Comment (djc) [inline src/acceptor.rs:7]: It looks like `ready()` isn't actually being used, so maybe we don't need this? (In which case we can lose the new dependency and thus the Cargo feature.)

**Review by djc — COMMENTED** [review]
Sorry, a few more thoughts, shaping up nicely though!

File: src/acceptor.rs (hunk 17 above)
Comment (Licenser) [reply src/acceptor.rs:7]: `ready!()` is used in a few places in the code below

File: src/acceptor.rs [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+mod builder;
+pub use builder::AcceptorBuilder;
+use builder::WantsTlsConfig;
+
+enum State {
+    Handshaking(tokio_rustls::Accept<AddrStream>),
+    Streaming(tokio_rustls::server::TlsStream<AddrStream>),
+}
+
+// tokio_rustls::server::TlsStream doesn't expose constructor methods,
```
[the remark this answers is not in the document]
Comment (Licenser) [reply src/acceptor.rs:24]: I've been taking a little look at this, and it seems quite tough to implement or require a partial rewrite. The TlsStream from the server example (this here) does the handshaking part which the TlsStream from `tokio-rustls` doesn't have, so that would change it (and I guess break other uses of it).

Perhaps it's best to tackle that in a different PR? it'll probably require a bit more discussion

[the remark this answers is not in the document]
Comment (djc) [reply src/acceptor.rs:24]: Ah sorry, sure, let's skip that!

**Comment by Licenser** [comment]
Okay second try, should be all done now :D

**Comment by Licenser** [comment]
The failing tests is why [this](https://github.com/rustls/hyper-rustls/pull/186#discussion_r1201809123) existed. Are you OK with putting that code back?

**Comment by djc** [comment]
Yup -- sorry, forgot about that gotcha.

File: src/acceptor.rs [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+// so we have to TlsAcceptor::accept and handshake to have access to it
+// TlsStream implements AsyncRead/AsyncWrite handshaking tokio_rustls::Accept first
```
Comment (cpu) [inline src/acceptor.rs:26]: I'm having trouble understanding the second line of this comment, on L26. Does this suggested edit track what you're trying to say? 

```suggestion
// TlsStream implements AsyncRead/AsyncWrite by handshaking with tokio_rustls::Accept first
```

File: src/acceptor.rs [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+    }
+}
+
+/// A TLS acceptor that can be used with hyper servers.
+pub struct TlsAcceptor {
+    config: Arc<ServerConfig>,
+    incoming: AddrIncoming,
+}
+
+/// A Acceptor for the `https` scheme.
```
Comment (cpu) [inline src/acceptor.rs:102]: grammar nit:
```suggestion
/// An Acceptor for the `https` scheme.
```

File: src/acceptor/builder.rs [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+    /// Configure ALPN accept protocols in order
+    pub fn with_alpn_protocols(
+        mut self,
+        alpn_protocols: Vec<Vec<u8>>,
+    ) -> AcceptorBuilder<WantsIncoming> {
+        self.0 .0.alpn_protocols = alpn_protocols;
+        AcceptorBuilder(WantsIncoming(self.0 .0))
+    }
+
+    /// Configure ALPN to accep HTTP/2
```
Comment (cpu) [inline src/acceptor/builder.rs:55]: typo:
```suggestion
    /// Configure ALPN to accept HTTP/2
```

**Review by cpu — COMMENTED** [review]
I'm not sure I'm familiar enough with the async ecosystem and Hyper in particular to give very meaningful feedback but I can surface a couple small nits :-)

**Comment by Licenser** [comment]
I may have found an alternative by updating the required features for the example, in theory :tm: (aka on my local machine) it should now not build during  unless the acceptor flag is present

**Comment by djc** [comment]
Nice, I was just looking for that this afternoon but couldn't quite find the documentation for it.

**Comment by Licenser** [comment]
I know you didn't like the `Default` implementation for the Builder, but clippy wants it and the CI is not passing. Do you prefer `Default` to return or would you prefer a `#[allow(...)]` for the function?

grafik

**Comment by Licenser** [comment]
Done, also added HTTP/1.0 support I think that's why windows / OS X were failing.

**Review by djc — APPROVED** [review]

**Comment by djc** [comment]
Thanks for working through all that!

**Comment by Licenser** [comment]
Woooh :D thanks!
